cub/device/dispatch/dispatch_histogram.cuh

```diff
@@ -237,79 +241,56 @@ struct DispatchHistogram
     // Scales samples to evenly-spaced bins
     struct ScaleTransform
     {
+        using CommonT = typename cuda::std::common_type<LevelT, SampleT>::type;
+
+        static constexpr bool PRECOMPUTE_RECIPROCAL =
+          cuda::std::is_floating_point<CommonT>::value;
+
         int    num_bins;    // Number of levels in array
-        LevelT max;         // Max sample level (exclusive)
-        LevelT min;         // Min sample level (inclusive)
-        LevelT scale;       // Bin scaling factor
+        CommonT max;        // Max sample level (exclusive)
+        CommonT min;        // Min sample level (inclusive)
+        CommonT scale;      // Bin scaling factor
 
         // Initializer
-        template <typename _LevelT>
         __host__ __device__ __forceinline__ void Init(
             int     num_output_levels,  // Number of levels in array
-            _LevelT max_,                // Max sample level (exclusive)
-            _LevelT min_,                // Min sample level (inclusive)
-            _LevelT scale_)              // Bin scaling factor
+            LevelT max_,                // Max sample level (exclusive)
+            LevelT min_)                // Min sample level (inclusive)
         {
             this->num_bins = num_output_levels - 1;
-            this->max = max_;
-            this->min = min_;
-            this->scale = scale_;
-        }
+            this->max = static_cast<CommonT>(max_);
+            this->min = static_cast<CommonT>(min_);
 
-        // Initializer (float specialization)
-        __host__ __device__ __forceinline__ void Init(
-            int    num_output_levels,   // Number of levels in array
-            float   max_,                // Max sample level (exclusive)
-            float   min_,                // Min sample level (inclusive)
-            float   scale_)              // Bin scaling factor
-        {
-            this->num_bins = num_output_levels - 1;
-            this->max = max_;
-            this->min = min_;
-            this->scale = float(1.0) / scale_;
-        }
-
-        // Initializer (double specialization)
-        __host__ __device__ __forceinline__ void Init(
-            int    num_output_levels,   // Number of levels in array
-            double max_,                 // Max sample level (exclusive)
-            double min_,                 // Min sample level (inclusive)
-            double scale_)               // Bin scaling factor
-        {
-            this->num_bins = num_output_levels - 1;
-            this->max = max_;
-            this->min = min_;
-            this->scale = double(1.0) / scale_;
+            // Optimize for floating point by precomputing the reciprocal:
+            CUB_IF_CONSTEXPR(PRECOMPUTE_RECIPROCAL)
+            {
+                this->scale = static_cast<CommonT>(num_output_levels - 1) /
+                              static_cast<CommonT>(max_ - min_);
```

Comment:
I'd rather do this arithmetic using the native types, and reusing the input variables reduces dependencies to improve ILP.

The compiler *should* optimize out the dependencies in this situation, but just in case...